### DISCUSSION TITLE
Laravels caching (artisan view:cache) breaks livewires intended caching for child components

### DISCUSSION THREAD

**Discussion by emil-nasso** [livewire/livewire#5658] [7. Bugs (archived)]
We are having problems with running nested livewire components on an app that runs on multiple server instances (with a load balancer in front).

This problem has been discussed before, and fixed, in this merge: https://github.com/livewire/livewire/pull/4354

However, there is one use-case that is not covered by that merge request, where the exact same problem still exists.

If I start my applications with an empty view cache, the relevant parts of the view cache will look like this:

Instance 1:

```php
if (! isset($_instance)) {
    $html = \Livewire\Livewire::mount('user-select')->html();
} elseif ($_instance->childHasBeenRendered('l3287825207-0')) {
    $componentId = $_instance->getRenderedChildComponentId('l3287825207-0');
    $componentTag = $_instance->getRenderedChildComponentTagName('l3287825207-0');
    $html = \Livewire\Livewire::dummyMount($componentId, $componentTag);
    $_instance->preserveRenderedChild('l3287825207-0');
} else {
    $response = \Livewire\Livewire::mount('user-select');
    $html = $response->html();
    $_instance->logRenderedChild('l3287825207-0', $response->id(), \Livewire\Livewire::getRootElementTagName($html));
}
``` 

Instance 2:

```php
if (! isset($_instance)) {
    $html = \Livewire\Livewire::mount('user-select')->html();
} elseif ($_instance->childHasBeenRendered('l3287825207-0')) {
    $componentId = $_instance->getRenderedChildComponentId('l3287825207-0');
    $componentTag = $_instance->getRenderedChildComponentTagName('l3287825207-0');
    $html = \Livewire\Livewire::dummyMount($componentId, $componentTag);
    $_instance->preserveRenderedChild('l3287825207-0');
} else {
    $response = \Livewire\Livewire::mount('user-select');
    $html = $response->html();
    $_instance->logRenderedChild('l3287825207-0', $response->id(), \Livewire\Livewire::getRootElementTagName($html));
}
```

As you can see, everything seems to be working like it should, and both instances uses the cache key `l3287825207-0`. In this state, the application behaves as expected.

However. If I run `php artisan view:cache`, it seems like the same code-path is not followed. I instead end up with the following in the view cache:

Instance 1:

```php
if (! isset($_instance)) {
    $html = \Livewire\Livewire::mount('user-select')->html();
} elseif ($_instance->childHasBeenRendered('cKiHams')) {
    $componentId = $_instance->getRenderedChildComponentId('cKiHams');
    $componentTag = $_instance->getRenderedChildComponentTagName('cKiHams');
    $html = \Livewire\Livewire::dummyMount($componentId, $componentTag);
    $_instance->preserveRenderedChild('cKiHams');
} else {
    $response = \Livewire\Livewire::mount('user-select');
    $html = $response->html();
    $_instance->logRenderedChild('cKiHams', $response->id(), \Livewire\Livewire::getRootElementTagName($html));
}
```

Instance 2:

```php
if (! isset($_instance)) {
    $html = \Livewire\Livewire::mount('user-select')->html();
} elseif ($_instance->childHasBeenRendered('nCtzgSo')) {
    $componentId = $_instance->getRenderedChildComponentId('nCtzgSo');
    $componentTag = $_instance->getRenderedChildComponentTagName('nCtzgSo');
    $html = \Livewire\Livewire::dummyMount($componentId, $componentTag);
    $_instance->preserveRenderedChild('nCtzgSo');
} else {
    $response = \Livewire\Livewire::mount('user-select');
    $html = $response->html();
    $_instance->logRenderedChild('nCtzgSo', $response->id(), \Livewire\Livewire::getRootElementTagName($html));
}
```

As seen above, the instances uses different cache-keys `cKiHams` and `nCtzgSo`.

The applications does not behave as expected in this state. Fields are "randomly" cleared when typing in them.

The problem seems to be that `LivewireManager::$currentCompilingViewPath` is always null when `artisan view:cache` generates the cache; https://github.com/livewire/livewire/blob/a6a52e78025d6b09b4962ec98f8e24f9de8d5955/src/LivewireBladeDirectives.php#L53

It seems like `RendersLivewireComponents::get` is never called when caching is done via `artican view:cache` but it is when the cache is generated via a live render:
https://github.com/livewire/livewire/blob/a6a52e78025d6b09b4962ec98f8e24f9de8d5955/src/ComponentConcerns/RendersLivewireComponents.php#L33

I'm not sure how this could be fixed, I assume it is not an easy fix, and the cache handing would have to be moved outside of the `RendersLivewireComponents::get` method.

Is adding a disclaimer about not using `php artisan view:cache` when using livewire to the documentation an option?

We run `php artisan view:cache` (and some other caching) when our applications start, so the workaround for us will probably be to stop doing that, for now, and rely on "live on render blade template caching", where it works nicely with livewire.

Please let me know if I can provide any more information.

**Comment by MattiaMarchiorato** [comment]
Same bug for us 👍

**Comment by bedo2991** [comment]
I just bumped into this in v. 3.4.12. Did anyone find a workaround?

**Comment by nuocgansoi** [comment]
Same bug with me.

**Comment by jackwh** [comment]
I'm also affected, from my investigations I think this runs a bit deeper when it comes to nested components:

- I have some nested components which are rendered conditionally
- After running `php artisan view:clear`, everything works great for the first request
- But after one page load (and the parent component becoming automatically cached), the nested components will repeatedly break until the next time `php artisan view:clear` is called.

There's a draft PR https://github.com/livewire/livewire/pull/8068 which may be relevant. Hope this can be fixed soon 🤞
